js/shutterbug-worker.js

```diff
@@ -121,7 +121,8 @@ export default class ShutterbugWorker {
 
   // Most important method. Returns HTML, CSS and dimensions of the snapshot.
   getHtmlFragment (callback) {
-    const $element = $(this.element)
+    const self = this
+    const $element = $(self.element)
```

Comment:
@kswenson Yeah, it didn't.

I think my note at the top isn't quite 100% accurate. I found in testing that when snapshots work, `this` in this context is the snapshot worker object, but when snapshots don't work, `this` is `window`. After making the change and some thorough testing it _seemed_ like it fixed the problem even though we didn't quite understand it ourselves. After we merged it in, however, we saw the problem occur again. So we'll undo this. 

The problem we're trying to fix is intermittent and seemingly somewhat random, unfortunately. It's difficult to tell when a change has an effect or not.

Reply:
The observation that `this` isn't bound correctly seems like an important one, but it suggests that the solution lies with the caller of the `getHtmlFragment()` function (or its caller, or its caller, ...) rather than within the function itself.

Reply:
@emcelroy I might have found it. There is a binding problem here:
https://github.com/concord-consortium/shutterbug.js/blob/584610d7fa9b5b8b03361388a30378ea782aa81b/js/shutterbug-worker.js#L68

The `_postMessageHandler` is not bound to this. 

But then in that _postMessageHandler:
https://github.com/concord-consortium/shutterbug.js/blob/ac4b4410f98e77eb0cf72fdd361a987d051e6225/js/shutterbug-worker.js#L246-L262

It seems to be expecting `this` to be bound correctly, because it is using `this._htmlFragRequestHandler.bind(this)`.

And in _htmlFragRequestHandler it is calling `this.getHtmlFragment`, so at that point `this` would probably be window.

However, I don't see how this ever worked, so I feel like I'm missing something.

Reply:
Oh but I just found:  `this._postMessageHandler = this._postMessageHandler.bind(this)` so it is being bound. So my message above is wrong.

But the complexity of all of this binding does make it seem possible for something to become bound wrong.

Reply:
Another thing to note is that this same method will be used in both the iframe and the host. So if your debugger is somehow combining them, then it could get confusing. In the iframe the element might really be the window.

Reply:
Thank you both for the thoughts. Regarding the element being the window in the iframe, I'm fairly certain what I was seeing was a `window` that was the Activity Player page. I was able to dig into the properties and found elements that were obviously part of AP, like the warning banner about my local AP storing data in a test database.

Noah had an idea that we haven't yet fully investigated about the request timeout settings in Shutterbug being the source of the problem. We're going to look more into that Thursday.

BTW: I did find, too, that if I changed the layout of my test activity so the snapshot target was always visible, the problem still occurred. So I think that rules out the earlier idea that the target was no longer in memory or otherwise "collapsed" somehow.

Reply:
Yes I didn't investigate the timeout path. I wouldn't be surprised if that triggers some kind of unbound call. Or possibly it just times out and the result is empty content. You could probably check that and change it to render a some text like "iframe timed out" instead of the iframe, so it would be easy to tell if that was happening.